key.core/src/main/java/de/uka/ilkd/key/rule/ObserverToUpdateRule.java

```diff
@@ -0,0 +1,428 @@
+ * {@code {U}[x = rec.query(params);]post} to
+ * {@code {U}{ x := query(heap, rec, params) }post} for model methods.
+ *
+ * For model fields, it turns
+ * {@code {U}[x = rec.query;]post} into
+ * {@code {U}{ x := query(heap, rec) }post}.
+ *
+ * Currently this supports model methods and model fields which can only be
+ * assigned to ghost entities.
+ *
+ * For strictly pure methods this is generally possible, but more side
+ * conditions would arise: Ensuring termination, strict purity, precondition, ...
+ *
+ * TODO Extend this mechanism beyond model methods to strictly pure methods.
+ *
+ * @author Mattias Ulbrich
+ */
+public final class ObserverToUpdateRule implements BuiltInRule {
+    /**
+     * The single instance.
+     */
+    public static final ObserverToUpdateRule INSTANCE = new ObserverToUpdateRule();
+
+    private static final Name NAME = new Name("Observer to update");
+
+    /**
+     * caching matching results
+     */
+    private static Term lastFocusTerm;
+    private static Union<Instantiation, ModelFieldInstantiation> lastInstantiation;
+
+    // -------------------------------------------------------------------------
+    // constructors
+    // -------------------------------------------------------------------------
+
+    private ObserverToUpdateRule() {
+        // private for singleton pattern
+    }
+
+    // -------------------------------------------------------------------------
+    // public interface
+    // -------------------------------------------------------------------------
+
+    @Override
+    public Name name() {
+        return NAME;
+    }
+
+    @Override
+    public String displayName() {
+        return NAME.toString();
+    }
+
+    @Override
+    public String toString() {
+        return displayName();
+    }
+
+
+    public IBuiltInRuleApp createApp(PosInOccurrence pos) {
+        return createApp(pos, null);
+    }
+
+    @Override
+    public IBuiltInRuleApp createApp(PosInOccurrence pos, TermServices services) {
+        return new DefaultBuiltInRuleApp(this, pos);
+    }
+
+    @Override
+    public boolean isApplicableOnSubTerms() {
+        return false;
+    }
+
+    @Override
+    public boolean isApplicable(Goal goal, PosInOccurrence pio) {
+        // focus must be top level succedent
+        if (pio == null || !pio.isTopLevel() || pio.isInAntec()) {
+            return false;
+        }
+
+        // abort if inside of transformer
+        if (Transformer.inTransformer(pio)) {
+            return false;
+        }
+
+        // instantiation must succeed
+        Union<Instantiation, ModelFieldInstantiation> inst =
+            instantiate(pio.subTerm(), goal.proof().getServices());
+        if (inst == null) {
+            return false;
+        }
+
+        if (inst.isFirst()) {
+            // additional checks for method calls.
+            // currently only applicable to strictly pure methods
+            // TODO check that this is not a two-state model method.
```

Comment:
What about this TODO? I could not figure out how to get this information here ...

Reply:
Was not so difficult to fix. Handled in 447578efb813ba350e7da92b93c10c3eae14b034.